cmd/tendermint/commands/util.go

```diff
@@ -0,0 +1,160 @@
+package commands
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+)
+
+// indexTaggedStructFields walks through a struct, looking up
+// those fields tagged by the requested tagName e.g:
+//    Foo int `json:"foo"`
+//    Bar string `mapstructure:"bar"`
+// recursively flattening anonymous struct fields by calling indexTaggedStructFields
+// on them.
+func indexTaggedStructFields(v interface{}, tagName string) map[string]interface{} {
+	val := reflect.ValueOf(v)
+	fieldsMap := make(map[string]interface{})
+	if val.Kind() == reflect.Ptr {
+		val = reflect.Indirect(val)
+	}
+
+	typ := val.Type()
+	n := val.NumField()
+
+	for i := 0; i < n; i++ {
+		fVal := val.Field(i)
+		field := typ.Field(i)
+		tag, _, squash := extractTag(tagName, field)
+		if tag != "" {
+			fieldsMap[tag] = fVal.Interface()
+		} else if squash {
+			innerMap := indexTaggedStructFields(fVal.Interface(), tagName)
+			// Since this is an anonymous struct,
+			// we should squash as requested it's content inside
+			for k, v := range innerMap {
+				if _, alreadyIn := fieldsMap[k]; !alreadyIn {
+					fieldsMap[k] = v
+				}
+			}
+		}
+	}
+
+	return fieldsMap
+}
+
+func extractTag(tagName string, f reflect.StructField) (theTagName string, omitEmpty, squash bool) {
+	tag := f.Tag.Get(tagName)
+	splits := strings.Split(tag, ",")
+	if len(splits) == 0 {
+		return "", false, false
+	}
+	if len(splits) > 1 {
+		switch splits[1] {
+		case "omitempty":
+			omitEmpty = true
+		case "squash":
+			squash = true
+		}
+	}
+	return splits[0], omitEmpty, squash
+}
+
+func isZero(v interface{}) bool {
+	switch v {
+	case false, nil, 0:
+		return true
+	default:
+		return false
+	}
+}
+
+// filterOutUnknownFields tallies fields
+// retrieved by parsing out fields in want
+// with those in gotFieldsMap and
+// reporting an error if there are fields in
+// gotFieldsMap that don't match up.
+//
+// A usecase is to report fields that were set in
+// the config file for a command but are unexpected.
+func filterOutUnknownFields(want interface{}, gotFieldsMap map[string]interface{}, tagName string, excusedFlags map[string]map[string]bool) error {
+	wantFieldsMap := indexTaggedStructFields(want, tagName)
```

Comment:
don't you think it will be better to extract this call and pass `wantFieldsMap` to `filterOutUnknownFields`?

```
func filterOutUnknownFields(wantFieldsMap map[string]interface{}, gotFieldsMap map[string]interface{}, excusedFlags map[string]map[string]bool) error {
```

Reply:
I could, however we still need to pass in the tagName for the internal caller inside https://github.com/orijtech/tendermint/blob/11f3c811332779f438f45322b433b399711ca748/cmd/tendermint/commands/util.go#L93.
If that weren't the case, I agree, I'd actually be nicer to extract it out.

Reply:
Can't we simplify this by having `indexTaggedStructFields` grab all sub structs and put them under keys like `p2p.laddr` ? Then we don't need to call `indexTaggedStructFields` in `filterOutUnknownFields` and don't care about vars like `specializedCommand`. Then we can do what Anton suggested too.